### PR TITLE
API: Restructure WSIReader to allow support for additional formats

### PR DESCRIPTION
**Pull request by shaneahmed**
- Restructure WSIReader as parent class to allow support to read whole slide images in other formats.
- Add OpenSlideWSIReader
- Add tests for OpenSlideWSIReader
- Add docs for OpenSlideWSIReader

### CODE REVIEW

File: tiatoolbox/dataloader/slide_info.py [hunk 1]
```diff
@@ -1,20 +1,21 @@
...
-    """Single file run to output or save WSI meta data.
-
-    Multiprocessing uses this function to run slide_info in parallel
+def slide_info(input_path, output_dir=None, verbose=True):
+    """Returns WSI meta data. Multiprocessing decorator runs this function in parallel.
 
     Args:
         input_path (str): Path to whole slide image
         output_dir (str): Path to output directory to save the output
+        verbose(bool): Print output, default=True
```
Comment (simongraham) [inline tiatoolbox/dataloader/slide_info.py:16]: space after verbose

File: tiatoolbox/dataloader/slide_info.py [hunk 2]
```diff
@@ -34,16 +35,18 @@ def slide_info(input_path, output_dir=None):
...
             input_dir=input_dir, file_name=file_name, output_dir=output_dir
         )
-        info = wsi_reader.slide_info()
+        info = wsi_reader.slide_info
+        if verbose:
+            print(info)
     else:
-        print("File type not supported")
-        info = None
+        raise FileNotSupported
```
Comment (simongraham) [inline tiatoolbox/dataloader/slide_info.py:50]: Doing it this way will raise a generic error - do we not want to also give the filename for which it failed on?

File: tiatoolbox/utils/exceptions.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+"""Custom Errors and Exceptions for TIAToolbox"""
+
+
+class FileNotSupported(Exception):
```
Comment (simongraham) [inline tiatoolbox/utils/exceptions.py:4]: instead - word as:  'File format is not supported'

File: tiatoolbox/dataloader/slide_info.py (hunk 1 above)
Comment (shaneahmed) [reply tiatoolbox/dataloader/slide_info.py]: Fixed in this commit https://github.com/TIA-Lab/tiatoolbox/commit/3ac21d1b573dc9f272a8527ef14e866b52c9c631

File: tiatoolbox/dataloader/slide_info.py (hunk 2 above)
Comment (shaneahmed) [reply tiatoolbox/dataloader/slide_info.py]: Updated it in this commit https://github.com/TIA-Lab/tiatoolbox/commit/3ac21d1b573dc9f272a8527ef14e866b52c9c631

File: tiatoolbox/utils/exceptions.py (hunk 3 above)
Comment (shaneahmed) [reply tiatoolbox/utils/exceptions.py:4]: Updated in this commit https://github.com/TIA-Lab/tiatoolbox/commit/dd194c4ee46e2bc4f36c73138f8f03515037b1fd

**Comment by shaneahmed** [comment]
> Overall the PR is good and I like how we subclass the WSIReader. The next step can be to subclass with a JP2Reader. Why do we use double underscore for `__slide_info__`? This is normally reserved for inbuilt python magic methods - we can instead just use double leading underscore to denote a private method. For the exceptions, do you think we should instead simply use a function, with the filename as an argument. This way we can give a more specific error message. Just something to consider.
> 
> Other than that, a couple of very minor things to fix before merging. I will approve, but please look into making those small changes before making the merge. Also, for the module string added to the top of a script, should we not add a space in between the module string and first library import?

Thanks Simon. I have addressed your concerns on ___slide_info__ and error message. Regarding space after docstring at the top of module, let's discuss this on the channel and then make it consistent after adding to OneNote.

File: tiatoolbox/cli.py [hunk 4]
```diff
@@ -119,22 +126,33 @@ def slide_info(wsi_input, output_dir, file_types, mode, workers=None):
 )
 def read_region(wsi_input, region, level, output_path, mode):
     """Reads a region in an whole slide image as specified"""
-    if all(region):
+    if not region:
```
Comment (John-P) [inline tiatoolbox/cli.py:129]: Is this checking if that region is not None? I would consider changing it to `if region is not None:` if so as it is more clear.

Comment (shaneahmed) [reply tiatoolbox/cli.py:129]: Its to check empty tuple "()".

**Pull request merged**